src/trie2.rs

```diff
@@ -0,0 +1,469 @@
+        for k in vec_in {
+            let k_hash = my_hash(&k);
+            vec.push((k,k_hash,()));
+        };
+        Trie{meta:Self::meta(vec.len()), 
+             rec:TrieRec::Leaf(TrieLeaf{kvs:Rc::new(vec)})}
+    }
+
+    fn split_bits (bits:&Bits) -> (Bits, Bits) {
+        let lbits = Bits{len:bits.len+1, bits:/* zero ------ */ bits.bits };
+        let rbits = Bits{len:bits.len+1, bits:(1 << bits.len) | bits.bits };
+        (lbits, rbits)
+    }
+
+    // TODO-Soon: Opt: After splitting a vec, create leaves by first checking whether the vec is empty.
+
+    fn leaf_or_empty (kvs:Vec<(K,HashVal,V)>) -> TrieRec<K,V> {
+        if kvs.len() == 0 { TrieRec::Empty }
+        else { TrieRec::Leaf(TrieLeaf{kvs:Rc::new(kvs)}) }
+    }
+
+    fn is_wf_rec (t:&TrieRec<K,V>, bits:Bits) -> bool {
+        match *t {
+            TrieRec::Empty => true,
+            TrieRec::Leaf(ref leaf) => {
+                // Check that all of the hash values match the given bit pattern of bits.
+                for &(_,ref hv, _) in leaf.kvs.iter() {
+                    for i in 0..bits.len {
+                        if ((bits.bits & (1 << i)) as usize) == (hv.0 & (1 << i)) { continue }
+                        else { return false }
+                    }
+                }
+                return true
+            }
+            // Check bit patterns match, and that recursive trees are well-formed.
+            TrieRec::Bin(ref b) => { 
+                let (b0, b1) = Self::split_bits(&bits);
+                let lwf = Self::is_wf_rec(&get!(b.left), b0);
+                let rwf = Self::is_wf_rec(&get!(b.right), b1);
+                b.bits == bits && lwf && rwf 
+            }
+        }
+    }
+
+    fn is_wf(self:&Self) -> bool {
+        Self::is_wf_rec(&self.rec, Bits{bits:0, len:0})
+    }
+
+    pub fn join (n:Option<Name>, lt: Self, rt: Self) -> Self {
+        //assert_eq!(lt.gauge, rt.gauge); // ??? -- Or take the min? Or the max? Or the average?
+        let gauge = if lt.meta.gauge > rt.meta.gauge { lt.meta.gauge } else { rt.meta.gauge };
+        Trie{rec:Self::join_rec(TrieMeta{gauge:gauge}, n, lt.rec, rt.rec, Bits{len:0, bits:0}),..lt}
+    }
+
+    fn join_rec (meta:TrieMeta, n:Option<Name>, lt: TrieRec<K,V>, rt: TrieRec<K,V>, bits:Bits) -> TrieRec<K,V> {
+        match (lt, rt) {
+            (TrieRec::Empty,   TrieRec::Empty)   => TrieRec::Empty,
+            (TrieRec::Empty,   TrieRec::Leaf(r)) => TrieRec::Leaf(r),
+            (TrieRec::Leaf(l), TrieRec::Empty  ) => TrieRec::Leaf(l),
+            (TrieRec::Leaf(l), TrieRec::Leaf(r)) => {
+                if l.kvs.len() == 0 { 
+                    TrieRec::Leaf(r)
+                } else if r.kvs.len() == 0 {
+                    TrieRec::Leaf(l)
+                } else if l.kvs.len() + r.kvs.len() < meta.gauge {
+                    // Sub-Case: the leaves, when combined, are smaller than the gauge.
+                    let mut vec = (*l.kvs).clone();
```

Comment:
Optimization: make a `Vec::with_capacity()` and then `extend()` it with the left side before pushing the right. You'll avoid resizing on push.